diff --git a/gnulib/import/glob.c b/gnulib/import/glob.c
index f1b20d4869..b10bea1d9f 100644
--- a/gnulib/import/glob.c
+++ b/gnulib/import/glob.c
@@ -21,7 +21,7 @@
    optimizes away the pattern == NULL test below.  */
 # define _GL_ARG_NONNULL(params)
 
-# include <config.h>
+# include <libc-config.h>
 
 #endif
 
diff --git a/gnulib/import/libc-config.h b/gnulib/import/libc-config.h
index 124f1d77e0..f1e9f59c81 100644
--- a/gnulib/import/libc-config.h
+++ b/gnulib/import/libc-config.h
@@ -181,3 +181,7 @@
 /* A substitute for glibc <shlib-compat.h>, good enough for Gnulib.  */
 #define SHLIB_COMPAT(lib, introduced, obsoleted) 0
 #define versioned_symbol(lib, local, symbol, version)
+
+#ifndef __THROWNL
+#define __THROWNL __THROW
+#endif
