diff --git a/bfd/elf-bfd.h b/bfd/elf-bfd.h
index 15767245..d7662e0f 100644
--- a/bfd/elf-bfd.h
+++ b/bfd/elf-bfd.h
@@ -3066,7 +3066,13 @@ static inline bfd_boolean
 bfd_section_is_ctf (const asection *sec)
 {
   const char *name = bfd_section_name (sec);
-  return strncmp (name, ".ctf", 4) == 0 && (name[4] == 0 || name[4] == '.');
+
+  return (
+    name[0] == '.'
+    && name[1] == 'c'
+    && name[2] == 't'
+    && name[3] == 'f'
+    && (name[4] == 0 || name[4] == '.'));
 }
 
 #ifdef __cplusplus
